File: ml/generate_dataset.py
"""
Generate expert-labeled HeatHits / IniTify training dataset.

Labels follow the provisional PAGASA-band + risk-factor rules documented in
ml/data/LABELING_GUIDE.md (same logic as src/config/decision-tree.rules.ts).

This is EXPERT-LABELED SYNTHETIC data for model development — not real survey data.
"""

from __future__ import annotations

import csv
import itertools
import random
from pathlib import Path

DATA_DIR = Path(__file__).parent / "data"
CSV_PATH = DATA_DIR / "heat_risk_dataset.csv"
XLSX_PATH = DATA_DIR / "heat_risk_dataset.xlsx"

HEADERS = [
    "record_id",
    "heat_index",
    "age",
    "health_condition",
    "activity_level",
    "hydration_status",
    "risk_level",
    "pagasa_band",
    "notes",
]

HEALTH_NONE = {"None", "Healthy", "N/A"}
ACTIVITY_LEVELS = ["Low", "Moderate", "High"]
HYDRATION_STATUSES = ["Well hydrated", "Moderately hydrated", "Dehydrated"]
HEALTH_CONDITIONS = [
    "None",
    "Healthy",
    "Hypertension",
    "Diabetes",
    "Asthma",
    "Heart disease",
]
RISK_LEVELS = ["LOW", "MODERATE", "HIGH", "EXTREME"]

PAGASA_BANDS = [
    (22, "Below caution (<27)"),
    (29, "Caution (27–32)"),
    (37, "Extreme caution (33–41)"),
    (46, "Danger (42–51)"),
    (54, "Extreme danger (≥52)"),
]


def pagasa_band(heat_index: float) -> str:
    if heat_index >= 52:
        return "Extreme danger (≥52)"
    if heat_index >= 42:
        return "Danger (42–51)"
    if heat_index >= 33:
        return "Extreme caution (33–41)"
    if heat_index >= 27:
        return "Caution (27–32)"
    return "Below caution (<27)"


def has_health_risk(health: str) -> bool:
    return health.strip().lower() not in {
        "none",
        "n/a",
        "no condition",
        "healthy",
        "",
    }


def count_risk_factors(age: int, health: str, activity: str, hydration: str) -> int:
    count = 0
    if age >= 60:
        count += 1
    if has_health_risk(health):
        count += 1
    if activity == "High":
        count += 1
    if hydration == "Dehydrated":
        count += 1
    return count


def label_risk(
    heat_index: float, age: int, health: str, activity: str, hydration: str
) -> str:
    """Expert labeling aligned with provisional IniTify decision tree."""
    if heat_index >= 42:
        return "EXTREME"

    factors = count_risk_factors(age, health, activity, hydration)

    if heat_index >= 33:
        if factors >= 1:
            return "EXTREME"
        return "HIGH"

    if heat_index >= 27:
        if factors >= 2:
            return "HIGH"
        if factors == 1:
            return "HIGH"
        return "MODERATE"

    # Below 27 °C
    if factors >= 1:
        return "MODERATE"
    return "LOW"


def build_rows() -> list[dict[str, str | int | float]]:
    rows: list[dict[str, str | int | float]] = []
    record_id = 1

    # Structured grid: heat bands × ages × representative profiles
    age_samples = [8, 16, 25, 35, 48, 55, 62, 71, 78]
    heat_samples = [22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 41, 43, 45, 48, 52, 55]

    profiles = [
        ("None", "Low", "Well hydrated", "Typical low-exposure profile"),
        ("None", "Moderate", "Moderately hydrated", "Average daily activity"),
        ("None", "High", "Moderately hydrated", "Outdoor worker, no comorbidity"),
        ("Hypertension", "Low", "Well hydrated", "Controlled condition, indoor"),
        ("Diabetes", "Moderate", "Moderately hydrated", "Chronic condition"),
        ("Asthma", "High", "Dehydrated", "High exposure + dehydration"),
        ("Heart disease", "Moderate", "Dehydrated", "Cardiac vulnerability"),
        ("None", "High", "Dehydrated", "Dehydrated active youth"),
        ("Healthy", "Low", "Dehydrated", "Dehydration without comorbidity"),
    ]

    seen: set[tuple] = set()

    for heat, age, (health, activity, hydration, note) in itertools.product(
        heat_samples, age_samples, profiles
    ):
        key = (heat, age, health, activity, hydration)
        if key in seen:
            continue
        seen.add(key)

        risk = label_risk(heat, age, health, activity, hydration)
        rows.append(
            {
                "record_id": f"HH-{record_id:03d}",
                "heat_index": heat,
                "age": age,
                "health_condition": health,
                "activity_level": activity,
                "hydration_status": hydration,
                "risk_level": risk,
                "pagasa_band": pagasa_band(heat),
                "notes": note,
            }
        )
        record_id += 1

    # Extra random diverse rows
    random.seed(42)
    while len(rows) < 120:
        heat = random.choice([23, 27, 31, 33, 39, 42, 44, 50, 53])
        age = random.randint(5, 85)
        health = random.choice(HEALTH_CONDITIONS)
        activity = random.choice(ACTIVITY_LEVELS)
        hydration = random.choice(HYDRATION_STATUSES)
        key = (heat, age, health, activity, hydration)
        if key in seen:
            continue
        seen.add(key)
        rows.append(
            {
                "record_id": f"HH-{record_id:03d}",
                "heat_index": heat,
                "age": age,
                "health_condition": health,
                "activity_level": activity,
                "hydration_status": hydration,
                "risk_level": label_risk(heat, age, health, activity, hydration),
                "pagasa_band": pagasa_band(heat),
                "notes": "Random synthetic profile",
            }
        )
        record_id += 1

    return rows


def write_csv(rows: list[dict]) -> None:
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    with CSV_PATH.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=HEADERS)
        writer.writeheader()
        writer.writerows(rows)
    print(f"Wrote {len(rows)} rows -> {CSV_PATH}")


def write_xlsx(rows: list[dict]) -> None:
    try:
        from openpyxl import Workbook
        from openpyxl.styles import Font, PatternFill
    except ImportError:
        print("openpyxl not installed — skipping .xlsx (CSV works in Excel)")
        return

    wb = Workbook()
    ws = wb.active
    ws.title = "Heat Risk Dataset"

    header_fill = PatternFill("solid", fgColor="1F4E79")
    header_font = Font(bold=True, color="FFFFFF")

    ws.append(HEADERS)
    for col in range(1, len(HEADERS) + 1):
        cell = ws.cell(row=1, column=col)
        cell.fill = header_fill
        cell.font = header_font

    for row in rows:
        ws.append([row[h] for h in HEADERS])

    ws.freeze_panes = "A2"
    ws.auto_filter.ref = f"A1:{chr(64 + len(HEADERS))}{len(rows) + 1}"

    # Column widths
    widths = [12, 12, 8, 18, 16, 22, 12, 28, 36]
    for i, width in enumerate(widths, start=1):
        ws.column_dimensions[chr(64 + i)].width = width

    wb.save(XLSX_PATH)
    print(f"Wrote {len(rows)} rows -> {XLSX_PATH}")


def main() -> None:
    rows = build_rows()
    write_csv(rows)
    write_xlsx(rows)

    counts: dict[str, int] = {}
    for row in rows:
        level = str(row["risk_level"])
        counts[level] = counts.get(level, 0) + 1
    print("Risk level distribution:", counts)


if __name__ == "__main__":
    main()

File: ml/generate_dataset.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DATA_DIR = path.join(__dirname, 'data');

const HEALTH = ['None', 'Healthy', 'Hypertension', 'Diabetes', 'Asthma', 'Heart disease'];
const ACTIVITY = ['Low', 'Moderate', 'High'];
const HYDRATION = ['Well hydrated', 'Moderately hydrated', 'Dehydrated'];

function pagasaBand(h) {
  if (h >= 52) return 'Extreme danger (>=52)';
  if (h >= 42) return 'Danger (42-51)';
  if (h >= 33) return 'Extreme caution (33-41)';
  if (h >= 27) return 'Caution (27-32)';
  return 'Below caution (<27)';
}

function hasHealthRisk(h) {
  return !['none', 'n/a', 'no condition', 'healthy', ''].includes(
    String(h).trim().toLowerCase(),
  );
}

function countFactors(age, health, activity, hydration) {
  let c = 0;
  if (age >= 60) c++;
  if (hasHealthRisk(health)) c++;
  if (activity === 'High') c++;
  if (hydration === 'Dehydrated') c++;
  return c;
}

function labelRisk(heat, age, health, activity, hydration) {
  if (heat >= 42) return 'EXTREME';
  const f = countFactors(age, health, activity, hydration);
  if (heat >= 33) return f >= 1 ? 'EXTREME' : 'HIGH';
  if (heat >= 27) return f >= 1 ? 'HIGH' : 'MODERATE';
  return f >= 1 ? 'MODERATE' : 'LOW';
}

const profiles = [
  ['None', 'Low', 'Well hydrated', 'Typical low-exposure profile'],
  ['None', 'Moderate', 'Moderately hydrated', 'Average daily activity'],
  ['None', 'High', 'Moderately hydrated', 'Outdoor worker no comorbidity'],
  ['Hypertension', 'Low', 'Well hydrated', 'Controlled condition indoor'],
  ['Diabetes', 'Moderate', 'Moderately hydrated', 'Chronic condition'],
  ['Asthma', 'High', 'Dehydrated', 'High exposure dehydration'],
  ['Heart disease', 'Moderate', 'Dehydrated', 'Cardiac vulnerability'],
  ['None', 'High', 'Dehydrated', 'Dehydrated active youth'],
  ['Healthy', 'Low', 'Dehydrated', 'Dehydration without comorbidity'],
];

const heats = [22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 41, 43, 45, 48, 52, 55];
const ages = [8, 16, 25, 35, 48, 55, 62, 71, 78];

const allRows = [];
const seen = new Set();
let id = 1;

for (const heat of heats) {
  for (const age of ages) {
    for (const [h, a, hy, n] of profiles) {
      const key = [heat, age, h, a, hy].join('|');
      if (seen.has(key)) continue;
      seen.add(key);
      allRows.push({
        record_id: `HH-${String(id++).padStart(3, '0')}`,
        heat_index: heat,
        age,
        health_condition: h,
        activity_level: a,
        hydration_status: hy,
        risk_level: labelRisk(heat, age, h, a, hy),
        pagasa_band: pagasaBand(heat),
        notes: n,
      });
    }
  }
}

// Pick a balanced subset (~120 rows) covering all bands and risk levels
const TARGET = 120;
const byRisk = {};
for (const r of allRows) {
  (byRisk[r.risk_level] ||= []).push(r);
}

function pickBalanced(pool, n) {
  const step = Math.max(1, Math.floor(pool.length / n));
  const out = [];
  for (let i = 0; i < pool.length && out.length < n; i += step) out.push(pool[i]);
  return out;
}

const perLevel = Math.floor(TARGET / 4);
const rows = [
  ...pickBalanced(byRisk.LOW || [], perLevel),
  ...pickBalanced(byRisk.MODERATE || [], perLevel),
  ...pickBalanced(byRisk.HIGH || [], perLevel),
  ...pickBalanced(byRisk.EXTREME || [], perLevel),
];

let seed = 42;
function rnd() {
  seed = (seed * 1103515245 + 12345) & 0x7fffffff;
  return seed / 0x7fffffff;
}

const rowKeys = new Set(rows.map((r) => [r.heat_index, r.age, r.health_condition, r.activity_level, r.hydration_status].join('|')));
id = rows.length + 1;

while (rows.length < TARGET) {
  const candidate = allRows[Math.floor(rnd() * allRows.length)];
  const key = [candidate.heat_index, candidate.age, candidate.health_condition, candidate.activity_level, candidate.hydration_status].join('|');
  if (rowKeys.has(key)) continue;
  rowKeys.add(key);
  rows.push({ ...candidate, record_id: `HH-${String(id++).padStart(3, '0')}` });
}

rows.sort((a, b) => a.record_id.localeCompare(b.record_id));
rows.forEach((r, i) => {
  r.record_id = `HH-${String(i + 1).padStart(3, '0')}`;
});

const headers = [
  'record_id',
  'heat_index',
  'age',
  'health_condition',
  'activity_level',
  'hydration_status',
  'risk_level',
  'pagasa_band',
  'notes',
];

const esc = (v) => `"${String(v).replace(/"/g, '""')}"`;
const csv = [headers.join(',')]
  .concat(rows.map((r) => headers.map((h) => esc(r[h])).join(',')))
  .join('\n');

fs.mkdirSync(DATA_DIR, { recursive: true });
fs.writeFileSync(path.join(DATA_DIR, 'heat_risk_dataset.csv'), csv, 'utf8');

try {
  const XLSX = (await import('xlsx')).default;
  const sheet = XLSX.utils.json_to_sheet(rows, { header: headers });
  const wb = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(wb, sheet, 'Heat Risk Dataset');
  XLSX.writeFile(wb, path.join(DATA_DIR, 'heat_risk_dataset.xlsx'));
  console.log('Wrote Excel -> ml/data/heat_risk_dataset.xlsx');
} catch {
  console.log('Excel skipped (CSV opens in Excel). Run: npm install xlsx');
}

const counts = {};
for (const r of rows) counts[r.risk_level] = (counts[r.risk_level] || 0) + 1;
console.log(`Wrote ${rows.length} rows -> ml/data/heat_risk_dataset.csv`);
console.log('Risk distribution:', counts);
